### ISSUE TITLE
Asyncio Driver Shouldn't Wrap CancelledError

### ISSUE THREAD

**Issue by mehaase** [rethinkdb/rethinkdb#6296]
The `connect()` method in the Asyncio driver has an exception handling clause like this:

    try:
        self._parent.handshake.reset()
        response = None
        with translate_timeout_errors():
            while True:
                request = self._parent.handshake.next_message(response)
                if request is None:
                    break
                # This may happen in the `V1_0` protocol where we send two requests as
                # an optimization, then need to read each separately
                if request is not "":
                    self._streamwriter.write(request)

                response = yield from asyncio.wait_for(
                    _read_until(self._streamreader, b'\0'),
                    timeout, loop=self._io_loop,
                )
                response = response[:-1]
    except ReqlAuthError:
        yield from self.close()
        raise
    except ReqlTimeoutError as err:
        yield from self.close()
        raise ReqlDriverError(
            'Connection interrupted during handshake with %s:%s. Error: %s' %
                (self._parent.host, self._parent.port, str(err)))
    except Exception as err:
        yield from self.close()
        raise ReqlDriverError('Could not connect to %s:%s. Error: %s' %
                              (self._parent.host, self._parent.port, str(err)))

The last `except` block is problematic, because if a task is canceled while it's suspended at `response = yield from asyncio.wait_for(...)`, that `asyncio.CancelledError` will get wrapped up in a `ReqlDriverError` with an empty error message. This is hard to debug (because of the empty error message), and it also breaks any cancel-friendly code. For example:

    try:
        await something_that_is_safe_to_cancel()
    except asyncio.CancelledError:
        pass
  
This says that if `something_that_is_safe_to_cancel()` is cancelled, we should just ignore it and move on. But if `something_that_is_safe_to_cancel()` calls RethinkDB's `connect()` method, then the raised exception is actually `ReqlDriverError`, not `CancelledError`. After a lengthy debugging session, I finally figured this out and wrote this workaround code:

    try:
        await something_that_is_safe_to_cancel()
    except r.ReqlDriverError as rde:
        if isinstance(rde.__context__, asyncio.CancelledError):
            pass
        else:
            raise

I consider this a bug in the RethinkDB driver, and my suggested fix is to add a penultimate `except` block:

    ...
    except asyncio.CancelledError:
      yield from self.close()
      raise
    except Exception as err:
        yield from self.close()
        raise ReqlDriverError('Could not connect to %s:%s. Error: %s' %
                              (self._parent.host, self._parent.port, str(err)))

I can submit a patchfile or PR if you agree. Thanks!

**Comment by AtnNn** [comment]
Your proposed patch sounds reasonable. If no one objects I would be happy to merge it.

I don't know enough about `asyncio` or the semantics of `CancelledError` to review this change in depth, but I'm pretty sure that a blanket `except Exception` in a library is a bad pattern.

**Comment by mehaase** [comment]
Paging @EggieCode and @dalanmiller... do you mind looking at this issue? I saw your PR for asynchronous iterators and so I think you understand this code much better than I do.

As I work more with the driver, I think my suggestion above is insufficient. I continue to have problems with the futures in the `_user_queries` dict throwing `InvalidStateError`, meaning somebody is trying to set a future result when the future is already done.

I suspect this is happening because each future is [awaited here](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/python/rethinkdb/asyncio_net/net_asyncio.py#L258), and if the task that is waiting on the future is canceled, then the future itself is also cancelled. Later on, the `_reader` coroutine will try [to set the future's result](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/python/rethinkdb/asyncio_net/net_asyncio.py#L283) which throws an exception. This exception is caught and handled by closing the connection, which then tries [to set the future's result again](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/python/rethinkdb/asyncio_net/net_asyncio.py#L283). A simple solution here might be to check in `close()` if the future is already done before calling `set_exception()`, but I'm still not sure if that would be a comprehensive solution. Thoughts?

**Comment by EggieCode** [comment]
The patch is reasonable. The only thing that bother me is why you would cancel that Connection in asyncio.
Is it that if you cancel the request query that hole connection breaks down?
Maybe add a check in_reader that check that the user_queries future is no canceld?

**Comment by mehaase** [comment]
I worked on reproducing this in the shortest amount of code possible. This bug doesn't seem to happen for select queries or for insert queries, but only happens for update queries...

```python
import asyncio
import rethinkdb as r

async def do_update(conn):
    await r.table('test').filter({'foo': 'bar'}) \
                         .update({'foo': 'baz'}).run(conn)


async def main():
    conn = await r.connect() # My connect() is different, but I don't think that matter

    print('Creating test data...')
    await r.table('test').delete().run(conn)
    docs = [{'n':n, 'foo':'bar'} for n in range(100)]
    await r.table('test').insert(docs).run(conn)

    print('Beginning update...')
    task = asyncio.ensure_future(do_update(conn))

    print('Cancelling update...')
    await asyncio.sleep(0)
    task.cancel()
    try:
        await task
    except asyncio.CancelledError:
        pass

    print('Closing connection...')
    await conn.close()

print('Loop starting...')
r.set_loop_type('asyncio')
loop = asyncio.get_event_loop()
loop.run_until_complete(main())
loop.close()
print('Loop closed')
```

In my real application, I have multiple tasks that are running various queries, and I occasionally need to kill some tasks. When I need to kill a task, I call its `cancel()` method and then await it. As this example shows, if that task is suspended inside a RethinkDB update query, then the query future is cancelled. When the connection is later closed, the driver tries to set the future result again, which leads to an exception:

```
root@738b0ef8acec:/starbelly# python3.6 test3.py 
Loop starting...
Creating test data...
Beginning update...
Cancelling update...
Closing connection...
Traceback (most recent call last):
  File "test3.py", line 24, in main
    await task
  File "/usr/lib/python3.6/asyncio/tasks.py", line 304, in _wakeup
    future.result()
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test3.py", line 34, in <module>
    loop.run_until_complete(main())
  File "/usr/lib/python3.6/asyncio/base_events.py", line 449, in run_until_complete
    return future.result()
  File "/usr/lib/python3.6/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test3.py", line 29, in main
    await conn.close()
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 105, in __next__
    return self.gen.send(None)
  File "/usr/local/lib/python3.6/dist-packages/rethinkdb/asyncio_net/net_asyncio.py", line 294, in close
    return (yield from ConnectionBase.close(self, *args, **kwargs))
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 105, in __next__
    return self.gen.send(None)
  File "/usr/local/lib/python3.6/dist-packages/rethinkdb/asyncio_net/net_asyncio.py", line 205, in close
    future.set_exception(ReqlDriverError(err_message))
asyncio.futures.InvalidStateError: invalid state
```

I changed `net_asyncio.py` (on version 2.3.0) as follows:

```
204         for query, future in iter(self._user_queries.values()):
205             if not future.done():
206                 future.set_exception(ReqlDriverError(err_message))
```

This fixes the `InvalidStateError` and now my demo runs fine. Can somebody try reproducing this and let me know if you get the same outcome? I haven't tried this on master.

**Comment by EggieCode** [comment]
Tested the name script and it crashes on python3.5. You shoud make the pull request for it to make asyncio more reliable.
